### ISSUE TITLE
Visual comparison / diff of texts

### ISSUE THREAD

**Issue by rufuspollock** [rgrp/cop21#9]
Want nice visual diffs between versions of the agreement text which we display online - probably in `diff/` directory:
- side by side
- **word** as well as paragraph diffing

Want to do this between different versions of the text - which we are now storing in different subfolders of `agreement`.

Diffs to do:
- The initial diff to try is between initial draft 23-oct-2015 and 4-dec-2015 version.
- Next would be final version `agreement/index.md` and e.g. 4-dec-2015 version (or a later one once we get it up - see e.g. #15 and #16)
### Implementation

Run a diff library on the texts and render to html and cache.

I've made a start trying out python difflib - see commit below. However, this is not good enough as per previous comment and we need to do better.

Qu: Could we get a hacky version by making sure we add texts in right order in github and then comparing between versions?

Ans: no - sadly. If we try doing an entire diff of agreement between e.g. draft and 4 dec it is "too much" and we don't get a diff, see e.g. https://github.com/okfn/cop21/compare/3f8a3395e80d879ae1a13c1f5ba917a3922b970c...cb1f967af1dd06fbc2d58fb03bc2367a8a95132b

**Comment by rufuspollock** [comment]
OK, i'm looking at a bunch of diff libraries.
- Python difflib: std html output just does paragraph diff and does not highlight word changes (like github)
- https://github.com/ymattw/coderev - kind of good but not sure it does word stuff that we want
  - that said cdiff https://github.com/ymattw/cdiff seems to do word highlighting in the terminal
- https://github.com/rtfpessoa/diff2html - closest thing to what we want as far as i can do though it seems to be entirely javascript ...
- https://github.com/cemerick/jsdifflib - did not have time to look at in much detail

**Comment by pauloborges** [comment]
@rgrp,

I started looking at this and I have two possible diff outputs so far:
- http://pauloborges.github.io/cop21/diff/diff1.html (using coderev/codediff)
- http://pauloborges.github.io/cop21/diff/diff2.html (using diff2html.py)

Tomorrow I'll try the diff2html javascript code.

What do you think?

**Comment by ghost** [comment]
@pauloborges, thanks for looking!
`diff2html.py` looks way more better than `coderev/codediff`:
- `diff2html.py` highlights all changes in each line which is really useful.
- `coderev/codediff` diff can't even fit into my viewport. 
- `coderev/codediff` splits large blocks of changes so comparing becomes painful.

**Comment by rufuspollock** [comment]
@pauloborges good work.

As @mark-in says the diff2 looks much better and personally i prefer python and no javascript so happy to stick with diff2html.py (though it might be nice to tweak css).

Do you want to have a go at putting up both the 4dec diff and the 4dec -> final version diff and submitting a pull request.

@mark-in you are very welcome to help here - please comment on any issues you are interested in helping with.

@tommv you may also want to take a look

**Comment by tommv** [comment]
I did take I look and diff2 looks indeed promising.
Someone also suggested me this software
http://www.juxtasoftware.org/
in case you want to try it our

**Comment by pauloborges** [comment]
I started tweaking a few css properties:
- http://pauloborges.github.io/cop21/diff/23-oct-vs-4-dec.html
- http://pauloborges.github.io/cop21/diff/4-dec-vs-12-dec.html

**Comment by ghost** [comment]
@pauloborges, thanks! Do you have a repo for this? **UPD**: yes, you are. I found it. :)
Why does 4-dec text is wider than 12-dec one [here](http://borges.pw/cop21/diff/4-dec-vs-12-dec.html)?

**Comment by ghost** [comment]
Just got an idea that it would be nice if each line would have an `id` parameter to be linkable via fragment identifier (e.g. `#n75`) like it does in `cgit`.

**Comment by pauloborges** [comment]
@mark-in,

Good question! I'm not very good with html/css so I don't have an answer yet. The fragment thing is a nice idea.

@rgrp,

I've tried to use the website layout, but I clearly have to improve the css even further. This'll take a while: http://borges.pw/cop21/diff/23-oct-vs-4-dec/

**Comment by rufuspollock** [comment]
@borges - this looks good already - can we get a pull request in and then work on improving the css.

**Comment by rufuspollock** [comment]
@pauloborges can you summarize here what is now live on the site and where as a result of your pull request (ps: would be have been good for the PR to have referenced this issue - something for the future ;-) ...) 

Also looking at: http://cop21.okfnlabs.org/diff/23-oct-vs-4-dec/ - is there any way to make the columns wider and to have paragraphs "top-aligned" rather than middle aligned?

**Comment by pauloborges** [comment]
PR #18 added the first version (`diff/*`) of the diff pages and the script (`scripts/diff.sh`) that created them.

We still have work to be done here, mostly related to presentation (CSS). I'll try to work on it but some help would be great since I'm not very comfortable with CSS.

**Comment by rufuspollock** [comment]
@pauloborges i'd seen the PR as i'd merged it ;-) I guess having a /diff/ index page that lists the diff we have would be useful. Also some info on the diff page as to what it is is important.

Also can we add a diff for new 9-dec text against the final text (in agreement/index.md).

**Comment by rufuspollock** [comment]
@pauloborges how are you doing here? Probably only 30m or so to complete this :-)

**Comment by pauloborges** [comment]
Hi! Give me until tomorrow and I should have this working.

**Comment by rufuspollock** [comment]
@pauloborges great :-) - how is it going?

**Comment by rufuspollock** [comment]
@pauloborges one last ping ;-) if you could push what you had that would be great!

**Comment by rufuspollock** [comment]
@GregoryRhysEvans suggest we add:
- 4 dec - 9 dec
- 9 dec - final text

And list these on front page in diff section (and in sidebar in a new diff section)

**Comment by rufuspollock** [comment]
@GregoryRhysEvans how are we doing here? I note that some texts sitll have header info in the diff e.g. http://cop21.okfnlabs.org/diff/9-dec-vs-12-dec/ - can you correct this?

Also how are you doing on sorting out the column widths in the diff?

**Comment by GregoryRhysEvans** [comment]
Fixed; any remaining issues will go in new issue thread.
